### PR TITLE
Domain: Allow setting a custom domain with Marketplace add-on.

### PR DESCRIPTION
**Pull request by retrofox**
#### Changes proposed in this Pull Request

This PR allows changing the primary domain of the site when it has the `set-primary-custom-domain` feature activated.

## Data approach

The most number files here belong to the site-features store integration. 

### `<QuerySiteFeatures />`

This component takes over requesting the site features hitting the WP COM API endpoint.

### Syncing data
Once the site features data is requests and the response is ready, it's processed and stored into the site state subtree.

### Allow setting the primary domain
The domains manager section checks whether the site has the `set-primary-custom-domain` featured activated in order to enable the UI to switch the primary domain.

### Running tests

** sites reducer**
yarn run test-client:watch client/state/sites/test

** site features reducer**
yarn run test-client:watch client/state/sites/features/test

** sites features selectors**
yarn run test-client:watch client/state/selectors/test/is-requesting-site-features.js
yarn run test-client:watch client/state/selectors/test/get-site-features.js
yarn run test-client:watch client/state/selectors/test/has-active-site-feature.js

#### Redux

You can take a look at the redux store using the dev tool when loading the plugins management page:

Screen Shot 2021-06-08 at 12 49 11 PM

#### Inspecting the global state object

Since the `features` is located in the site state subtree, you can assed to it using the console just by typing `state.sites.features`. You can inspected the populated tree:

image

#### Inspecting the List component

Under the hood, a new `canSetPrimaryDomain` property is populated in the `List` component. You could take a look at it there in case you'd like to check it.

Screen Shot 2021-06-08 at 12 52 48 PM

### Testing with UI

* ~You need to apply  D59668-code changes in case they haven't landed yet.~
* You need to apply D61913-code changes in case they haven't landed yet.
* Get a site with a **Free** plan and **Yoast Premium product**
* Go to domains manage page http://calypso.localhost:3000/domains/manage/<your-wac-with-free-and-yoast-site>
* Get an additional domain if the site has only one
* Try to change the primary domain

before | after
----|------
Screen Shot 2021-05-25 at 8 44 16 AM | Screen Shot 2021-05-25 at 8 35 44 AM

* Repeat the tests but this time with a site with a Free plan without the Yoast premium plan. You shouldn't be able to change the primary domain.

### video

https://user-images.githubusercontent.com/77539/119491136-f5d08f80-bd33-11eb-881e-8bddcacf55c7.mov

Related to #

### CODE REVIEW

File: packages/calypso-products/src/constants/jetpack.ts [hunk 1]
```diff
@@ -23,6 +23,7 @@ export const PRODUCT_JETPACK_CRM = 'jetpack_crm';
 export const PRODUCT_JETPACK_CRM_MONTHLY = 'jetpack_crm_monthly';
 export const PRODUCT_JETPACK_CRM_FREE = 'jetpack_crm_free';
 export const PRODUCT_JETPACK_CRM_FREE_MONTHLY = 'jetpack_crm_free_monthly';
+export const PRODUCT_YOAST_PREMIUM = 'yoast_premium';
```
Comment (allilevine) [inline packages/calypso-products/src/constants/jetpack.ts]: Why is this a Jetpack constant?

**Review by allilevine — COMMENTED** [review]
Looks good! I was able to add a domain to my free site with Marketplace add-on and change it to be the primary domain:
Screen Shot 2021-05-25 at 4 30 40 PM

On a free site without a Marketplace add-on, I can't change a custom domain to be a primary domain:
Screen Shot 2021-05-25 at 4 42 32 PM

I just had a question, below.

File: packages/calypso-products/src/constants/jetpack.ts (hunk 1 above)
Comment (retrofox) [reply packages/calypso-products/src/constants/jetpack.ts]: why not? 😅 changed here https://github.com/Automattic/wp-calypso/pull/53126/commits/39d07164130b2c27837395389b4c8198ade9d6bc

**Review by pablinos — COMMENTED** [review]
This works as expected in conjunction with D61913-code, but as discussed, I wonder if we should remove this logic from the front end and extend one of the API endpoints to return a flag based on the server-side version of this logic.

**Comment by retrofox** [comment]
> This works as expected in conjunction with D61913-code, but as discussed, I wonder if we should remove this logic from the front end and extend one of the API endpoints to return a flag based on the server-side version of this logic.

✔️ 

Now, the approach checks whether the site has the `set primary custom domain` feature. If so, it will allow setting the custom domain as the primary of the site. Naturally, you'll need to apply the patch D61913-code

File: client/state/sites/features/actions.js [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+		return new Promise( ( resolve ) => {
+			wpcom.undocumented().getSiteFeatures( siteId, ( error, data ) => {
+				if ( error ) {
+					debug( 'Fetching site features failed: ', error );
+
+					const errorMessage =
+						error.message ||
+						i18n.translate(
+							'There was a problem fetching site features. Please try again later or contact support.'
```
Comment (a8ci18n) [inline client/state/sites/features/actions.js:56]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **21** times:
`translate( 'There was a problem fetching site products. Please try again later or contact support.' )` *ES Score: 7*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/yoast-premium-allow-setting-primary-domain)

File: client/lib/wpcom-undocumented/lib/undocumented.js [hunk 3]
```diff
@@ -952,6 +952,25 @@ Undocumented.prototype.getSitePlans = function ( siteDomain, fn ) {
...
+ * @param {Function} fn The callback function
+ */
+Undocumented.prototype.getSiteFeatures = function ( siteDomain, fn ) {
+	debug( '/sites/:site_domain:/features query' );
+
+	return this._sendRequest(
+		{
+			path: `/sites/${ encodeURIComponent( siteDomain ) }/features`,
+			method: 'get',
+			apiVersion: '1.1',
```
Comment (pablinos) [inline client/lib/wpcom-undocumented/lib/undocumented.js:897]: Is there a v2 endpoint that will get us this information?

File: client/my-sites/domains/domain-management/list/index.jsx [hunk 4]
```diff
@@ -344,7 +347,12 @@ export class List extends React.Component {
 	};
 
 	shouldUpgradeToMakeDomainPrimary( domain ) {
-		const { isDomainOnly, isOnFreePlan, hasNonPrimaryDomainsFlag } = this.props;
+		const {
+			isDomainOnly,
+			isOnFreePlan,
```
Comment (pablinos) [inline client/my-sites/domains/domain-management/list/index.jsx:352]: Do we need to check this now? If they're on a paid plan then they should have the feature to set the primary domain.

Similarly would do we need to check for the 'domain only' plan still?

File: client/state/sites/features/README.md [hunk 5]
```diff
@@ -0,0 +1,38 @@
+# Site Features
+
+A module for managing site features data.
```
Comment (pablinos) [inline client/state/sites/features/README.md]: Should we elaborate a little on what the features are, and what they can be used for?

File: packages/calypso-products/src/constants/features.ts [hunk 6]
```diff
@@ -19,6 +19,8 @@ import {
 export const FEATURE_WP_SUBDOMAIN = 'wordpress-subdomain';
 export const FEATURE_BLOG_DOMAIN = 'blog-domain';
 export const FEATURE_CUSTOM_DOMAIN = 'custom-domain';
+export const FEATURE_CUSTOM_DOT_BLOG_DOMAIN = 'custom-dot-blog-domain';
```
Comment (pablinos) [inline packages/calypso-products/src/constants/features.ts]: Why are we adding this constant/feature with this change? It's related to the product feature this supports, but should it be added with the rest of the .blog domain product work?

File: client/my-sites/domains/domain-management/list/index.jsx [hunk 7]
```diff
@@ -463,6 +472,7 @@ export class List extends React.Component {
 
 		return [
 			<QuerySitePurchases key="query-purchases" siteId={ selectedSite.ID } />,
```
Comment (pablinos) [inline client/my-sites/domains/domain-management/list/index.jsx:474]: Could we bring back the active features as part of this purchases query? I'm curious if we could get away with a single API call.

**Review by pablinos — COMMENTED** [review]
This works well for me. I just had a few thoughts.

Should we try and get an external review? I'm not 100% on the current best practice with new bits of the state tree, but this looks fine to me.

File: client/lib/wpcom-undocumented/lib/undocumented.js (hunk 3 above)
Comment (retrofox) [reply client/lib/wpcom-undocumented/lib/undocumented.js:897]: no, there isnt'. It's `v1`

File: client/my-sites/domains/domain-management/list/index.jsx (hunk 4 above)
Comment (retrofox) [reply client/my-sites/domains/domain-management/list/index.jsx:352]: 100% agree, but I'd like to hold on removing it for now. I suggest creating a new follow-up PR.

File: packages/calypso-products/src/constants/features.ts (hunk 6 above)
Comment (retrofox) [reply packages/calypso-products/src/constants/features.ts]: sorry I'm going to remove this constant.

File: client/state/sites/features/README.md (hunk 5 above)
Comment (retrofox) [reply client/state/sites/features/README.md]: Yes, for sure. Could we do that in a follow-up PR?

File: client/my-sites/domains/domain-management/list/index.jsx (hunk 7 above)
Comment (retrofox) [reply client/my-sites/domains/domain-management/list/index.jsx:474]: sorry, are you suggesting using the same query component for both requests?

File: client/state/sites/features/README.md (hunk 5 above)
Comment (retrofox) [reply client/state/sites/features/README.md]: Docummenting now

File: client/my-sites/domains/domain-management/list/index.jsx (hunk 4 above)
Comment (pablinos) [reply client/my-sites/domains/domain-management/list/index.jsx:352]: Ok let's do that. If we can, maybe we can add a test, so we can convince ourselves that the logic hasn't changed

File: client/my-sites/domains/domain-management/list/index.jsx (hunk 7 above)
Comment (pablinos) [reply client/my-sites/domains/domain-management/list/index.jsx:474]: I was wondering if that would be possible. We're interested in the subscriptions that have been purchased for the site, and this retrieves that from the API. I was wondering if we could extend it with the features data.

It's probably a fairly big change to the approach, and I'm not sure if it's possible or makes sense, so we can leave it and come back to it later if necessary.

**Review by pablinos — APPROVED** [review]
I've left the one conversation unresolved to remind us to create an issue for the follow-up PR. Apart from that, let's ship this once the associated patch is merged

File: client/my-sites/domains/domain-management/list/index.jsx (hunk 4 above)
Comment (retrofox) [reply client/my-sites/domains/domain-management/list/index.jsx:352]: sounds very good to me.

**Comment by retrofox** [comment]
> I've left the one conversation unresolved to remind us to create an issue for the follow-up PR.

Sounds good to me 👍 

> Apart from that, let's ship this once the associated patch is merged

This is not mandatory. The `/features` endpoint is already implemented in prod so I think we're safe merging den deploying these changes. Having said this, it's totally fine waiting for the patch too.

**Comment by pablinos** [comment]
> The /features endpoint is already implemented in prod

Ah of course, and we're not using this feature in production yet. It feels slightly safer to merge the patch first, but feel free to do it the other way around.

**Pull request merged**